Address config changes and new defaults for Solr 8.

kmcelwee opened the pull request.

rlskoeser commented on parasolr/solr/tests/test_schema.py:
```diff
@@ -26,19 +26,19 @@ def test_delete_field(self, test_client):
 
         test_client.schema.add_field(name='A', type='string')
-        test_client.schema.replace_field(name='A', type='int')
+        test_client.schema.replace_field(name='A', type='pint')
```
Can you provide a link to help me understand this change? I'm not finding much when I search for it.

kmcelwee replied:
This caused me a few headaches, so I'll try to break it down:
* `int` in Solr 6's `basic_configs` references the `TrieIntField`.
     * https://github.com/apache/lucene-solr/blob/6eb7bc3b7bd6dea2a104448e921f7234973f0197/solr/server/solr/configsets/_default/conf/managed-schema#L213
* Solr documentation says that `TrieIntField` is deprecated: “Use `IntPointField` instead.” 
     * https://lucene.apache.org/solr/guide/8_3/field-types-included-with-solr.html
* `IntPointField` in `_default` is named `pint`.
     * https://github.com/apache/lucene-solr/blob/6eb7bc3b7bd6dea2a104448e921f7234973f0197/solr/server/solr/configsets/_default/conf/managed-schema#L213
* It threw an error because `int` is no longer a name for a `fieldType` in the new configset `_default`.
     * (Searching `<fieldType name="int"` on [the schema](https://github.com/apache/lucene-solr/blob/6eb7bc3b7bd6dea2a104448e921f7234973f0197/solr/server/solr/configsets/_default/conf/managed-schema) brings up no results

**Most importantly: `pint` exists in Solr 6. So this ~should~ works for Solr 6 and Solr 8**

Edit: confirmed that it works on both versions.

rlskoeser replied:
thank you for explaining! what a pain

rlskoeser commented on .travis.yml:
```diff
@@ -6,7 +6,7 @@ jdk:
   - openjdk8 # recommended jdk for solr < 7
 env:
   global:
-    - SOLR_VERSION=6.6.5
```
Would like to see this switched to a build matrix so we can test against both.

rlskoeser commented on README.rst:
```diff
@@ -75,7 +75,7 @@ To use with Django:
         'COLLECTION': 'name',
         # any configSet in SOLR_ROOT/server/solr/configsets
-        'CONFIGSET': 'basic_configs' # optional, basic_configs is default
+        'CONFIGSET': '_default' # optional, _default is default
```
Please revise to mention both versions in the comment, e.g. something like:
`'CONFIGSET': '_default'  # optional, _default is default as of Solr x; basic_configs is default for Solr 6.x`

Do we know when this default changed? (8.x or 7.x)?

thatbudakguy replied:
i think it changed with 7, fwiw

thatbudakguy replied:
yep, see https://github.com/Princeton-CDH/parasolr/issues/48

rlskoeser commented on ci/testsettings.py:
```diff
@@ -30,7 +30,7 @@
             # aggressive commitWithin for test only
             'COMMITWITHIN': 750,
-            'CONFIGSET': 'basic_configs'
+            'CONFIGSET': '_default'
```
For the build matrix with my proposed approach, we would need to leave this off but figure out how to specify solr version based on the environment variable.

rlskoeser reviewed (commented):
The changes seem trivial enough that I would like to support both solr 6 and 8 in the same version of parasolr, so that we don't have to worry about maintaining two different versions. (Especially since I have a bunch of unreleased features I need for PPA, which is still Solr 6).

Here's what I suggest:
- solr config dictionary takes an optional solr version parameter; default is solr 6.x, so that existing applications don't have to make any changes
- make default configset dependent on solr version (maybe preset variables that we choose depending on versions)
- where the tests differ, just check the configured Solr version to figure out which type or name to use

kmcelwee commented:
I think we can now straddle the two versions. I got all test configs based off `testsettings.py`, and `testsettings.py` bases everything off the `SOLR_VERSION` environment variable. If `SOLR_VERSION` is not defined, it defaults to 8. 

Through this whole process, though, we've mostly just changed our *testing* scripts. A project (like Geniza or PPA) shouldn't need to update anything when Parasolr goes to 6.0 🎉

The biggest things I'd like your thoughts on:
* Does the logic in `testsettings.py` look clean to you? Do we want logic in settings?
* I removed `CONFIGSET` as an optional parameter because if it was not defined it would default to `_default`. (Note: this is the only thing I changed outside of our testing scripts, and would lead to breaking changes.)
* I dropped Python 3.5 just so that the tests ran faster (otherwise we'd be doubling Travis's run time). Should I add it back?

thatbudakguy commented on .travis.yml:
```diff
@@ -1,19 +1,24 @@
-    - DJANGO=2.0
-    - DJANGO=2.1
-    - DJANGO=2.2
+    # solr 8
+    - SOLR_VERSION=8.6.2 DJANGO=
+    - SOLR_VERSION=8.6.2 DJANGO=1.11
+    - SOLR_VERSION=8.6.2 DJANGO=2.0
+    - SOLR_VERSION=8.6.2 DJANGO=2.1
+    - SOLR_VERSION=8.6.2 DJANGO=2.2
+    # solr 6
+    - SOLR_VERSION=6.6.5 DJANGO=
+    - SOLR_VERSION=6.6.5 DJANGO=1.11
+    - SOLR_VERSION=6.6.5 DJANGO=2.0
+    - SOLR_VERSION=6.6.5 DJANGO=2.1
+    - SOLR_VERSION=6.6.5 DJANGO=2.2
```
is there a reason we don't test against 7?

rlskoeser replied:
We're not planning to use it ourselves, so why expand the matrix and slow the tests down? We're not aware of anyone using it, and hopefully someone new to parasolr would start with solr 8. We could always add it to our testing matrix if someone requests (and makes the case).

thatbudakguy replied:
wouldn't it add close to zero time if the matrix is running in parallel? seems odd to say we support it if we don't test for it. I guess we could say "supports solr>6; tested on solr 6 and 8"

thatbudakguy replied:
i guess we don't actually say "supports" in the README at all, which is smart. so maybe this isn't an issue :)

thatbudakguy commented on .travis.yml:
```diff
@@ -37,8 +42,6 @@ script:
 after_success:
 - codecov
 - if [[ "$TRAVIS_PULL_REQUEST" == "false" && "$TRAVIS_PYTHON_VERSION" == "3.6" ]]; then ./cc-test-reporter after-build --exit-code $TRAVIS_TEST_RESULT; fi
```
is the latter part of this conditional necessary if we're always using 3.6?

thatbudakguy commented on README.rst:
```diff
@@ -50,7 +50,7 @@ configuration and indexing content.
     :alt: Requirements Status
 
-Currently tested against Python 3.5 and 3.6, Solr 6.6.5, and Django 1.11,
+Currently tested against Python 3.5 and 3.6, Solr 6.6.5 and 8.6.2, and Django 1.11,
```
can remove the bit about python 3.5 here also

rlskoeser commented:
We can't drop python 3.5 yet because mep is still 3.5 (pretty sure) and our combined qa environment is 3.5. As soon as we get those upgraded we can drop it.

thatbudakguy commented:
oh - don't we need it back in the `.travis.yml` file then?

rlskoeser commented on README.rst:
```diff
@@ -120,6 +122,12 @@ requires a minimal settings file for Django-required configurations.
    cp ci/testsettings.py.sample testsettings.py
    python -c "import uuid; print('\nSECRET_KEY = \'%s\'' % uuid.uuid4())" >> testsettings.py
 
+- By default, parasolr expects Solr 8. If running tests with an earlier 
+  version of Solr, either explicitly change **MAJOR_SOLR_VERSION** in your
+  local **testsettings.py** or set the environment variable::
+
+   export SOLR_VERSION=x.x.x
+
```
I was probably unclear in my suggestions about how to implement this, sorry.

I'd like to see the Solr version as an optional config setting in the `SOLR_CONNECTIONS` dict. 

For the tests on travis-ci, my expectation is that the python config would need to be populated from an environment variable since we can't hard-code it in test settings.

rlskoeser replied:
... but maybe this doesn't actually matter, if the only changes are specific to how we test?

kmcelwee replied:
From conversation during pair: keeping this for now since it's only for testing and easily changeable wihtin the local `testsettings.py`

rlskoeser commented on ci/testsettings.py:
```diff
@@ -1,3 +1,7 @@
+import os
+
+MAJOR_SOLR_VERSION = int(os.environ.get('SOLR_VERSION', '8').split('.')[0])
+configset = 'basic_configs' if MAJOR_SOLR_VERSION < 7 else '_default'
```
I think this needs to be in parasolr code somewhere and not just in our test settings — aren't there are a few places besides tests where we need to know the default?

rlskoeser replied:
Or does it not matter because we're assuming solr 8, and applications that need to should override if necessary? (that sounds great)

kmcelwee replied:
From conversation during pairing: Reintroduce the default configset in non-test code.

rlskoeser commented on .travis.yml:
```diff
@@ -1,19 +1,24 @@
-    - DJANGO=2.2
+    # solr 8
+    - SOLR_VERSION=8.6.2 DJANGO=
+    - SOLR_VERSION=8.6.2 DJANGO=1.11
+    - SOLR_VERSION=8.6.2 DJANGO=2.0
+    - SOLR_VERSION=8.6.2 DJANGO=2.1
+    - SOLR_VERSION=8.6.2 DJANGO=2.2
```
mep and ppa are both on django 2.2 now, and versions prior to 2.2 are no longer supported by django — let's go ahead and drop them. (We can do that in a separate PR + chore if it makes more sense to you!)

rlskoeser reviewed (commented):
I was initially confused about how this version of parasolr would work against existing projects that are still running on solr 6 without requiring changes — but I'd forgotten (until I looked) that we're using custom configsets for those projects, so the default shouldn't matter (I'm not sure about pemm, but ok with not updating that one, we're not going to be adding any new features).  @kmcelwee am I understanding correctly?

It does seem preferable to default to solr 8 if we can make it work without breaking old projects.
